tests/text_input_v2_with_input_method_v1.cpp

```diff
@@ -0,0 +1,285 @@
+          text_input_manager{app_client.bind_if_supported<zwp_text_input_manager_v2>(wlcs::AnyVersion)},
+          text_input{zwp_text_input_manager_v2_get_text_input(text_input_manager, app_client.seat())},
+          input_method{input_client.bind_if_supported<zwp_input_method_v1>(wlcs::AnyVersion)}
+    {
+        zwp_input_method_v1_add_listener(input_method, &listener, this);
+    }
+
+    void create_focused_surface()
+    {
+        app_surface.emplace(app_client.create_visible_surface(100, 100));
+        the_server().move_surface_to(app_surface.value(), 0, 0);
+        pointer.move_to(10, 10);
+        pointer.left_click();
+        app_client.roundtrip();
+    }
+
+    void activate(zwp_input_method_context_v1* context)
+    {
+        input_method_context = std::make_unique<NiceMock<wlcs::MockInputMethodContextV1>>(context);
+    }
+
+    MOCK_METHOD1(deactivate, void(zwp_input_method_context_v1*));
+
+    static zwp_input_method_v1_listener constexpr listener {
+        wlcs::method_event_impl<&TextInputV2WithInputMethodV1Test::activate>,
+        wlcs::method_event_impl<&TextInputV2WithInputMethodV1Test::deactivate>
+    };
+
+    void enable_text_input()
+    {
+        create_focused_surface();
+        zwp_text_input_v2_enable(text_input, app_surface->wl_surface());
+        zwp_text_input_v2_update_state(text_input, 0, 0);
+        app_client.roundtrip();
+        input_client.roundtrip();
+    }
+
+    wlcs::Pointer pointer;
+    wlcs::Client app_client;
+    wlcs::Client input_client;
+    wlcs::WlHandle<zwp_text_input_manager_v2> text_input_manager;
+    NiceMock<wlcs::MockTextInputV2> text_input;
+    wlcs::WlHandle<zwp_input_method_v1> input_method;
+    std::unique_ptr<NiceMock<wlcs::MockInputMethodContextV1>> input_method_context;
+    std::optional<wlcs::Surface> app_surface;
+};
+
+TEST_F(TextInputV2WithInputMethodV1Test, text_input_enters_surface_on_focus)
+{
+    wl_surface* entered{nullptr};
+    EXPECT_CALL(text_input, enter(_, _))
+        .WillOnce(SaveArg<1>(&entered));
+    create_focused_surface();
+    EXPECT_THAT(entered, Eq(app_surface.value().wl_surface()));
+}
+
+TEST_F(TextInputV2WithInputMethodV1Test, text_input_activates_context_on_enable)
+{
+    create_focused_surface();
+    zwp_text_input_v2_enable(text_input, app_surface->wl_surface());
+    zwp_text_input_v2_update_state(text_input, 0, 0);
+    app_client.roundtrip();
+    input_client.roundtrip();
+    EXPECT_TRUE(input_method_context != nullptr);
+}
+
+TEST_F(TextInputV2WithInputMethodV1Test, text_input_deactivates_context_on_disable)
+{
+    create_focused_surface();
+
+    EXPECT_CALL(*this, deactivate(_));
+    zwp_text_input_v2_enable(text_input, app_surface->wl_surface());
+    zwp_text_input_v2_update_state(text_input, 0, 0);
+    app_client.roundtrip();
+    input_client.roundtrip();
+    zwp_text_input_v2_disable(text_input, app_surface->wl_surface());
+    zwp_text_input_v2_update_state(text_input, 1, 0);
```

Comment:
Again, I'm reasonably confident that `enable`/`disable` does not require committing with `update_state`.